src/drivers/imu/mpu6000/MPU6000.cpp

```diff
@@ -0,0 +1,989 @@
+	uint8_t data[sizeof(MPUReport)] {};
+	_interface->read(MPU6000_SET_SPEED(MPUREG_INT_STATUS, MPU6000_LOW_BUS_SPEED), data, sizeof(data));
+	PX4_WARN("error triggered");
+	print_registers();
+	_in_factory_test = false;
+}
+
+uint8_t
+MPU6000::read_reg(unsigned reg, uint32_t speed)
+{
+	uint8_t buf{};
+	_interface->read(MPU6000_SET_SPEED(reg, speed), &buf, 1);
+	return buf;
+}
+
+uint16_t
+MPU6000::read_reg16(unsigned reg)
+{
+	uint8_t buf[2] {};
+
+	// general register transfer at low clock speed
+	_interface->read(MPU6000_LOW_SPEED_OP(reg), &buf, arraySize(buf));
+	return (uint16_t)(buf[0] << 8) | buf[1];
+}
+
+int
+MPU6000::write_reg(unsigned reg, uint8_t value)
+{
+	// general register transfer at low clock speed
+	return _interface->write(MPU6000_LOW_SPEED_OP(reg), &value, 1);
+}
+
+void
+MPU6000::modify_reg(unsigned reg, uint8_t clearbits, uint8_t setbits)
+{
+	uint8_t	val = read_reg(reg);
+	val &= ~clearbits;
+	val |= setbits;
+	write_reg(reg, val);
+}
+
+void
+MPU6000::write_checked_reg(unsigned reg, uint8_t value)
+{
+	write_reg(reg, value);
+
+	for (uint8_t i = 0; i < MPU6000_NUM_CHECKED_REGISTERS; i++) {
+		if (reg == _checked_registers[i]) {
+			_checked_values[i] = value;
+		}
+	}
+}
+
+int
+MPU6000::set_accel_range(unsigned max_g_in)
+{
+	// workaround for bugged versions of MPU6k (rev C)
+	if (is_mpu_device()) {
+		switch (_product) {
+		case MPU6000ES_REV_C4:
+		case MPU6000ES_REV_C5:
+		case MPU6000_REV_C4:
+		case MPU6000_REV_C5:
+			write_checked_reg(MPUREG_ACCEL_CONFIG, 1 << 3);
+			_px4_accel.set_scale(CONSTANTS_ONE_G / 4096.0f);
+			return OK;
+		}
+	}
+
+	uint8_t afs_sel;
+	float lsb_per_g;
+	//float max_accel_g;
+
+	if (max_g_in > 8) { // 16g - AFS_SEL = 3
+		afs_sel = 3;
+		lsb_per_g = 2048;
+		//max_accel_g = 16;
+
+	} else if (max_g_in > 4) { //  8g - AFS_SEL = 2
+		afs_sel = 2;
+		lsb_per_g = 4096;
+		//max_accel_g = 8;
+
+	} else if (max_g_in > 2) { //  4g - AFS_SEL = 1
+		afs_sel = 1;
+		lsb_per_g = 8192;
+		//max_accel_g = 4;
+
+	} else {                //  2g - AFS_SEL = 0
+		afs_sel = 0;
+		lsb_per_g = 16384;
+		//max_accel_g = 2;
+	}
+
+	write_checked_reg(MPUREG_ACCEL_CONFIG, afs_sel << 3);
+
+	_px4_accel.set_scale(CONSTANTS_ONE_G / lsb_per_g);
+
+	return OK;
+}
+
+void
+MPU6000::start()
+{
+	/* make sure we are stopped first */
+	uint32_t last_call_interval = _call_interval;
+	stop();
+	_call_interval = last_call_interval;
+
+	if (!is_i2c()) {
+
+		_call.period = _call_interval - MPU6000_TIMER_REDUCTION;
+
+		/* start polling at the specified rate */
+		hrt_call_every(&_call,
+			       1000,
+			       _call_interval - MPU6000_TIMER_REDUCTION,
+			       (hrt_callout)&MPU6000::measure_trampoline, this);
+
+	} else {
+#ifdef USE_I2C
+		/* schedule a cycle to start things */
+		work_queue(HPWORK, &_work, (worker_t)&MPU6000::cycle_trampoline, this, 1);
+#endif
+	}
+}
+
+void
+MPU6000::stop()
+{
+	if (!is_i2c()) {
+		hrt_cancel(&_call);
+
+	} else {
+#ifdef USE_I2C
```

Comment:
Eliminating this kinda crap too would be nice. I'm going to end my review here, there is a lot of cruft in this driver and probably not worth trying to clean up and then regression test it.